### PR TITLE
Add Gitea Oauth

### PR DESCRIPTION
**Pull request by techknowlogick**
With fallback to basic auth if Gitea server does not support oauth

Linked to https://github.com/drone/go-login/pull/3

### CODE REVIEW

**Comment by bradrydzewski** [comment]
awesome!  one question: how does gitea handle the scenario where two separate go routines try to refresh the access token at the same time?  this is something that caused issues in early gitlab oauth implementations, so just want to understand a bit more :)

**Comment by techknowlogick** [comment]
Tests fail because the above linked PR isn't merged

**Comment by bradrydzewski** [comment]
I just merged :)

**Comment by techknowlogick** [comment]
Thanks @bradrydzewski 😄

**Comment by techknowlogick** [comment]
@jonasfranz could you answer @bradrydzewski's question above?

**Comment by techknowlogick** [comment]
Marking this as WIP as a patch needs to be sent to Gitea so that if no redirect_uri is set then it'll default to what is set.

**Comment by techknowlogick** [comment]
@bradrydzewski looking through Gitea OAuth code, refreshing token isn't implemented in Gitea (yet). I was able to compile drone and login with this PR and it pulled all the repos. I've also been able to submit an updated PR to go-login with being able to use RedirectURL.

File: cmd/drone-server/config/config.go [hunk 1]
```diff
@@ -293,9 +293,11 @@ type (
 
 	// Gitea provides the gitea client configuration.
 	Gitea struct {
-		Server     string `envconfig:"DRONE_GITEA_SERVER"`
-		SkipVerify bool   `envconfig:"DRONE_GITEA_SKIP_VERIFY"`
-		Debug      bool   `envconfig:"DRONE_GITEA_DEBUG"`
+		ClientID     string `envconfig:"DRONE_GITEA_CLIENT_ID"`
```
Comment (appleboy) [inline cmd/drone-server/config/config.go]:
```go
	Gitea struct {
		Server       string   `envconfig:"DRONE_GITEA_SERVER" default:"https://try.gitea.io"`
		ClientID     string   `envconfig:"DRONE_GITEA_CLIENT_ID"`
		ClientSecret string   `envconfig:"DRONE_GITEA_CLIENT_SECRET"`
		SkipVerify   bool     `envconfig:"DRONE_GITEA_SKIP_VERIFY"`
		Scope        []string `envconfig:"DRONE_GITEA_SCOPE" default:"repo,repo:status,user:email,read:org"`
		Debug        bool     `envconfig:"DRONE_GITEA_DEBUG"`
	}
```

File: cmd/drone-server/inject_login.go [hunk 2]
```diff
@@ -91,6 +92,16 @@ func provideGiteaLogin(config config.Config) login.Middleware {
...
 		return nil
 	}
+	if config.Gitea.ClientID != "" {
+		return &gitea.Config {
+			ClientID:     config.Gitea.ClientID,
+			ClientSecret: config.Gitea.ClientSecret,
+			Server:       config.Gitea.Server,
+			Client:       defaultClient(config.Gitea.SkipVerify),
+			Logger:       logrus.StandardLogger(),
+			RedirectURL:  config.Server.Addr + "/login",
```
Comment (appleboy) [inline cmd/drone-server/inject_login.go]: Add

```go
Scope:        config.Gitea.Scope,
```

File: cmd/drone-server/config/config.go (hunk 1 above)
Comment (techknowlogick) [reply cmd/drone-server/config/config.go]: Thanks for feedback 😄 I've made this change.

**Comment by appleboy** [comment]
also update the `provideClient` func:

```diff
    case config.Github.ClientID != "":
        return provideGithubClient(config)
-    case config.Gitea.Server != "":
+    case config.Gitea.ClientID != "":
        return provideGiteaClient(config)
    case config.GitLab.ClientID != "":
        return provideGitlabClient(config)
```

**Comment by techknowlogick** [comment]
@appleboy I kept `config.Gitea.ClientID != ""` codition so this doesn't break for Gitea users < 1.8 so they can still use username/password

File: cmd/drone-server/inject_login.go (hunk 2 above)
Comment (techknowlogick) [reply cmd/drone-server/inject_login.go]: done

**Comment by appleboy** [comment]
@techknowlogick Good catch. Thanks for the PR. LGTM.

**Comment by jonasfranz** [comment]
Refresh times are implemented. You can only use an access token or refresh token once to redeem a new token.

**Comment by bradrydzewski** [comment]
For context, GitLab had a similar restriction a few years ago that they since removed. Here are some issues that we ran into https://github.com/drone/drone/issues/855

File: cmd/drone-server/config/config.go [hunk 3]
```diff
@@ -293,9 +293,12 @@ type (
...
 	// Gitea provides the gitea client configuration.
 	Gitea struct {
-		Server     string `envconfig:"DRONE_GITEA_SERVER"`
-		SkipVerify bool   `envconfig:"DRONE_GITEA_SKIP_VERIFY"`
-		Debug      bool   `envconfig:"DRONE_GITEA_DEBUG"`
+		Server       string   `envconfig:"DRONE_GITEA_SERVER" default:"https://try.gitea.io"`
+		ClientID     string   `envconfig:"DRONE_GITEA_CLIENT_ID"`
+		ClientSecret string   `envconfig:"DRONE_GITEA_CLIENT_SECRET"`
+		SkipVerify   bool     `envconfig:"DRONE_GITEA_SKIP_VERIFY"`
+		Scope        []string `envconfig:"DRONE_GITEA_SCOPE" default:"repo,repo:status,user:email,read:org"`
```
Comment (jonasfranz) [inline cmd/drone-server/config/config.go]: Gitea does not support scopes currently

Comment (techknowlogick) [reply cmd/drone-server/config/config.go]: I decided that a documentation change is better than a code change for when scope is supported

**Comment by tboerger** [comment]
Great to see that this gets finally integrated

File: cmd/drone-server/config/config.go [hunk 4]
```diff
@@ -293,9 +293,12 @@ type (
 
 	// Gitea provides the gitea client configuration.
 	Gitea struct {
-		Server     string `envconfig:"DRONE_GITEA_SERVER"`
-		SkipVerify bool   `envconfig:"DRONE_GITEA_SKIP_VERIFY"`
-		Debug      bool   `envconfig:"DRONE_GITEA_DEBUG"`
+		Server       string   `envconfig:"DRONE_GITEA_SERVER" default:"https://try.gitea.io"`
```
Comment (bradrydzewski) [inline cmd/drone-server/config/config.go]: this default needs to be removed. An empty string is the only way we know if gitea is configured or not.

Comment (techknowlogick) [reply cmd/drone-server/config/config.go]: Good catch. Thanks :)

I've updated.

**Comment by appleboy** [comment]
@techknowlogick conflicts.

**Comment by jonasfranz** [comment]
@bradrydzewski do you propose to remove refresh token invalidation after a new token is signed?

**Comment by bradrydzewski** [comment]
@jonasfranz yes, I think ideally invalidating existing refresh tokens after a new refresh token is generated would be opt-in.  Most oauth providers (GitHub, Google, GitLab, etc) only invalidate refresh tokens if the user revokes access to the application.

**Comment by lunny** [comment]
@techknowlogick `go.mod` conflicted

**Comment by jonasfranz** [comment]
@bradrydzewski With my latest PR, Gitea will not invalidate refresh tokens by default. But there is an option to enable token invalidation.

**Comment by bradrydzewski** [comment]
@jonasfranz awesome, can you link to the pull request here. Once it is merged, we can merge this pull request. Also I will need to document the minimum version of Gitea that will be required -- can you let me know which version this will be? Thanks!!

**Comment by jonasfranz** [comment]
@bradrydzewski The PR is already merged: https://github.com/go-gitea/gitea/pull/6584

It will be in version 1.8.0 (or 1.8.0-rc3 if we have a rc3).

**Comment by techknowlogick** [comment]
@lunny conflict resolved.

**Comment by appleboy** [comment]
@bradrydzewski Already merged in Gitea.

**Comment by bradrydzewski** [comment]
excellent. one last request -- can we get an entry for this in the changelog? thanks!

**Comment by lunny** [comment]
@bradrydzewski I think yes. This could be in release notes of v1.8.

**Comment by bradrydzewski** [comment]
what I mean is that we need an entry for this change in Drone's changelog file https://github.com/drone/drone/blob/master/CHANGELOG.md

**Comment by techknowlogick** [comment]
@bradrydzewski added

**Comment by tboerger** [comment]
This works with gitea master? Or also with some stable release?

**Comment by techknowlogick** [comment]
@tboerger it works with 1.8.0 which was just released yesterday, but Drone can still work with older versions of gitea using basic auth.

**Comment by oliver006** [comment]
I've been using it successfully with docker images gitea:v1.8.0-rc3 (and 1.8.0 now that it's released) and drone/drone@sha256:7acfbba60ac6c35537b099ce612aac0af3cdbfa2e48d0ee96730cd4c904b89fa (the sha of the image right after this PR got merged).

But there seems to be a little problem with refreshing the gitea token. After a while, I'm no longer to e.g. enable a gitea repo for drone. I'm still logged in and can modify settings within drone. Once I log out and back in, everything is fine.

Should I open a separate issue?

**Comment by techknowlogick** [comment]
@oliver006 between rc3 and final gitea released a patch that should fix the issue you described above. Please feel free to open an issue in gitea tracker if this comes up again.

**Comment by oliver006** [comment]
Thanks for the quick response @techknowlogick . I tried with Gitea 1.8.0 and the drone/drone image I mentioned above but it's still not working. I get:

image

I checked in the DB, the `user_oauth_expiry` value is unchanged (and 2h in the past).

I'll open an issue in the Gitea project.

**Comment by bradrydzewski** [comment]
I think the issue is that if Gitea uses a refresh token, a refresher needs to be configured. See https://github.com/drone/drone/blob/master/cmd/drone-server/inject_login.go#L166

**Comment by frostieDE** [comment]
This features needs some documentation, or am I missing something on the docs? :smile:

**Comment by ukos-git** [comment]
here are my quick notes for installing:

- go to your gitea instance application settings page: `https://gitea.io/user/settings/applications`
- add a oauth instance with arbitrary id and redirect to the login page: `https://gitea.io/login`

Screenshot from 2019-04-28 15-40-02

- you will get ClientID and ClientSecret:

Screenshot from 2019-04-28 15-41-13

- add the generated values to your drone environement:

```console
DRONE_GITEA_CLIENT_ID=XXX
DRONE_GITEA_CLIENT_SECRET=XXX
```

be sure to use latest `gitea/gitea` and `drone/drone` images (gitea version is displayed in ui bottom left. Today is Gitea Version: 345bc06)

Probably not enough for a PR to docs but better than no documentation.

**Pull request merged**